src/base/emu-i386/simx86/protmode.c

```diff
@@ -390,3 +391,35 @@ int emu_ldt_write(unsigned char *paddr, uint32_t op, int len)
 	return 1;
 }
 
+void emu_check_read_pagefault(dosaddr_t addr, int len)
+{
+	int attr;
+	if (addr + len > LOWMEM_SIZE + HMASIZE &&
+	    !is_hardware_ram(addr, len) &&
+	    (attr = dpmi_get_page_attribute(addr, len)) == 0) {
+		/* trigger an exception in DPMI */
+		TheCPU.err = EXCP0E_PAGE;
+		/* uncommitted page is never "present" */
+		TheCPU.scp_err = 4;
+		TheCPU.cr2 = addr;
+		longjmp(jmp_env, 0);
+	}
+}
+
+int emu_check_write_pagefault(dosaddr_t addr, uint32_t op, int len)
+{
+	int attr;
+	if (addr + len > LOWMEM_SIZE + HMASIZE &&
+	    !is_hardware_ram(addr, len) &&
+	    (attr = dpmi_get_page_attribute(addr, len)) < 9) {
+		if (emu_ldt_write(MEM_BASE32(addr), op, len))
+			return 1;
+		/* trigger an exception in DPMI, either a readonly
+		   committed/shared page (1/2) or uncommitted page (0) */
+		TheCPU.err = EXCP0E_PAGE;
+		TheCPU.scp_err = 6 + ((attr & 7) != 0);
+		TheCPU.cr2 = addr;
```

Comment:
But is this correct?
It seems, addr is the beginning of the block,
whereas attr is the "most restrictive attribute",
but they may not match of you cross the page
boundaries?

Reply:
Hmm how does a real spanning read or write behave if the first byte is on a PROT_READ page and the second is on PROT_NONE?  I was assuming it would take on the properies of PROT_NONE (as both should fail) but I'll verify.

Reply:
My point was different.
Say, you have some long string insn
that crosses the page boundaries.
The most restrictive attr (say, PROT_NONE)
may be on the second page, but the fault
may happen on the first page simply because
it is "restrictive enough" (say, PROT_READ).

Reply:
So you may set the correct addr but the
attr from "most restrictive" page.

Reply:
string instructions don't call this yet (if ever...) you need something more complex for that, if you want to treat them at a higher level.

At the moment they just do things like this:
```
                        while (i--) { write_dword(dest, read_dword(src));
                            dest += 4; src += 4; }
```

I suppose that to speed those up you could do this page-by-page as in `memcpy_to_dos`, and don't let the checks span pages, but the function is now only thought to handle size 1,2,4 and handle the nasty page-spanning misaligned r/w's of those sizes.

Reply:
> 1,2,4 and handle the nasty page-spanning misaligned r/w's of those sizes.

So if eg size 4 spans 2 pages, and we
fault at first but the most restrictive is second,
then this func will just be called 2 times,
one for each page?

Reply:
No, if size=4, it's a dword read/write, then this func will only be called once, with the fault having the cr2 equal to the address (on the first page) and the fault error code corresponding to permissions on the second page.

Reply:
And why is this correct?

Reply:
it's not correct, it seems such accesses are not atomic: I tried (using KVM-vm86) to write a word to `9fff:000f` in vga mode 12h (planar mode) and it gave me a page fault with error code 6 and cr2=0xa0000, after which dosemu2 crashed.

Maybe should just split such accesses into multiple read_byte()/write_byte()s which also means I no longer need to find the most restrictive attribute, as it's guaranteed to be on a single page then.

Reply:
> means I no longer need to find the most restrictive attribute

I think "most restrictive" never makes sense.
Current attr for a particular address is a more
sensible API, but its already there. So indeed
splitting looks like the best approach. Maybe
you don't need to split the insn itself, but the
check can be splitted into 2. If both succeed,
then you do the entire write.

Reply:
I split the reads and writes in the end.